extensions/2.0/Khronos/KHR_materials_diffuse_transmission/README.md

```diff
@@ -0,0 +1,241 @@
+|![](figures/plant_1.0.jpg)<br><sub>`diffuseTransmissionFactor: : 1.0`<br>`diffuseTransmissionColorFactor: [1,0,0]`</sub> |
+
+<sub>[Potted Plant](https://polyhaven.com/a/potted_plant_02) by [Rico Cilliers](https://www.artstation.com/ricocilliers) on [PolyHaven](https://polyhaven.com)</sub>
+
+Increasing the `diffuseTransmissionFactor` takes energy from the diffuse reflection and transfers it to the diffuse transmission. In the images above this effect is clearly visible. The green appearance of the leaves is the product of front-lighting and diffuse reflections based on the `baseColorTexture`. With `diffuseTransmissionFactor` approaching `1.0`, the appearance changes to red, as a product of back-lighting and the `diffuseTransmissionColorFactor`. Specular reflections are untouched by this effect.
+
+## Implementation
+
+*This section is normative.*
+
+This extension changes the `dielectric_brdf` defined in [Appendix B](../../../../specification/2.0/README.md#appendix-b-brdf-implementation)
+```
+dielectric_brdf =
+  fresnel_mix(
+    ior = 1.5,
+    base = diffuse_brdf(color = baseColor),
+    layer = specular_brdf(α = roughness ^ 2)
+  )
+```
+ to the following:
+```
+dielectric_brdf =
+  fresnel_mix(
+    ior = 1.5,
+    base = mix(
+      diffuse_brdf(color = baseColor),
+      diffuse_btdf(color = diffuseTransmissionColor),
+      diffuseTransmission),
+    layer = specular_brdf(α = roughness ^ 2)
+  )
+```
+Increasing the strength of the diffuse transmission effect using the `diffuseTransmissionFactor` parameter takes away energy from the diffuse reflection BSDF and passes it to the diffuse transmission BSDF. The specular reflection BSDF and Fresnel weighting are not affected.
+
+*This section is non-normative.*
+
+With a simple Lambert BRDF model, `diffuse_brdf` and `diffuse_btdf` may be implemented as follows
+```
+function diffuse_brdf(color) {
+  if (view and light on same hemisphere) {
+    return (1/pi) * color
+  } else {
+    return 0
+  }
+}
+
+function diffuse_btdf(color) {
+  if (view and light on opposite hemispheres) {
+    return (1/pi) * color
+  } else {
+    return 0
+  }
+}
+
+function mix(bsdf0, bsdf1, factor) {
+  return (1-factor) * bsdf0 + factor * bsdf1
+}
+```
+
+<figure style="text-align:center">
+<img src="./figures/bsdf.svg"/>
+<sub><figcaption><em>Left: Diffuse BRDF. Right: Diffuse BTDF.</em></figcaption></sub>
+</figure>
+
+## Combining Diffuse Transmission with other Extensions
+### KHR_materials_transmission
+Both `KHR_materials_diffuse_transmission` and `KHR_materials_transmission` replace the diffuse BRDF with a mix of diffuse BRDF and a BTDF that transmits light onto the opposite side of the surface. In case of `KHR_materials_transmission`, this is a microfacet BTDF that shares its roughness with the microfacet BRDF. In case of `KHR_materials_diffuse_transmission`, on the other hand, this is a diffuse BTDF.
```

Comment:
I don't think it makes sense to combine the two extensions KHR_materials_transmission and KHR_materials_diffuse_transmission.
They do both model the same physical phenomena - ie, the transmission of light through the material.
They both have a transmission factor - which is the amount of light that enters into the material.
Where the original transmission extension can adapt the roughness, this value will scatter the light being transmitted - the new extension are for materials where no light will pass through without being scattered/refracted.
Using the original extension with roughness = 1 should produce the same result as this extension.

I see this extension as a great way of simplifying/optimizing shader performance since it is known that the background will never be visible through the material.

Only difference that I can see is that this extension has a parameter to set the color of the light passing through the material. 
However this should be possible by combining KHR_materials_transmission with the volume extension.

Reply:
This draft was written by Tobias and Bastien some years ago, and they are describing how the effect works in Enterprise PBR which is being emulated here.  That system is designed for path tracing, where it makes sense to offer both specular transmission (the original glTF transmission) and diffuse transmission (this PR).

I agree that we should strongly discourage realtime systems from applying both at once.

Reply:
>I agree that we should strongly discourage realtime systems from applying both at once.

I think it's more than that - they model different and non-compatible materials.
* This extension models the behavior under full/very high roughness - so high that you never see anything through the material.
* KHR_materials_transmission using varying values for roughness - where most values means that the material is (mostly) transparent.

Just because Enterprise PBR uses the effect together is not, in my mind, a reason for us to do exactly the same.

My view is that the materials that these extension model are not compatible and we should disalow the use of the extensions together.

Reply:
> This extension models the behavior under full/very high roughness

Actually no, this is a common misconception.  The surface of KHR_materials_diffuse_transmission is allowed to be smooth, and will produce specular reflections according to surface roughness as usual, including very shiny ones.  The transmitted light is diffuse, fully scattered, but not because of roughness at the surface.

Both this and the original KHR_materials_transmission extension are permitted to be present, and this section of the spec describes the interaction in mathematical detail.  Admittedly such details are primarily of use to path-tracing systems for which this math was originally worked out, and is not much use in a real-time rasterizer.

Perhaps we could add a note discouraging realtime assets from using both together.  But it's far too late to deeply alter such a fundamental aspect of this spec. This draft has been around for four years and has already shipped with stable versions of BabylonJS, for example.  We're not making breaking changes to it at this point, the time for that is long gone.  We're just looking to polish up the wording and get it ratified soon.

Reply:
>Actually no, this is a common misconception. The surface of KHR_materials_diffuse_transmission is allowed to be

My bad, I was hoping that you would understand what I meant - let me phrase it this way:
This extension models the behavior of full/high roughness of the inside (of the material)
All light that enters into the material is scattered - there can be no translucensy.
It is simply not compatible with KHR_materials_transmission from a physical point of view!

>Perhaps we could add a note discouraging realtime assets from using both together. But it's far too late to deeply alter such a fundamental aspect of this spec. This draft has been around for four years and has already shipped with stable

I really do not understand this statement!
We must seek to release clearly defined extensions - as of now this extension is not very well defined from a physical point of view. Regardless of current implementations or time.

> gone. We're just looking to polish up the wording and get it ratified soon.

Well, extensions shall be ratified when ready - this extension needs more work!

Reply:
As the co-author, I may be biased, but I'd say that this extension is very well defined in the context of the current material model and other ratified material extensions. :) There's a clear definition of how it interacts with KHR_materials_transmission (that should also work for rasterizers). You find identical or very similar definitions in other PBR models, i.e. OpenPBR.  

> ...as of now this extension is not very well defined from a physical point of view

What does that even mean? PBR appearance models do not necessarily replicate physics directly. Instead, they create abstract mathematical concepts that enable us to simulate the appearance of the real world in a computationally feasible manner, while adhering to important physical principles such as energy conservation. So, it is always a balancing act between physical correctness, performance, visual quality, artistic control and also inherently non-physical annoyances as infinitely thin surfaces.

Reply:
@bsdorra I have a related question about how the light breaks down.  Let's say at a particular point along the surface you have 10% Fresnel reflection, and you have transmission and diffuse transmission both set to 50% each on the same material.  This is how I think it breaks down, is this correct?

1. Fresnel / specular reflection takes some light off the top.  Of the original 100% received at the chosen point, 10% is reflected by this, and 90% is not yet accounted for.
2. Transmission (the specular kind) asks for 50%, but it gets 50% of what's left (90%), so for this point only 45% of the original received light shines through the surface along the specular transmission ray.  The remaining 45% of the original is not yet accounted for.
3. Diffuse transmission also asks for 50% of the remaining 45%.  As a result, 22.5% of the original light transmits through the surface and is released as a diffuse lobe on the opposite side.  An additional 22.5% of the original light is all that remains not yet accounted for.
4. All remaining light (22.5% of the original) is modulated by the base color and released as a diffuse lobe on the same side of the surface where it was received.

Does that sound right?

Reply:
@emackey that sounds correct to me 👍🏼

Reply:
>What does that even mean? PBR appearance models do not necessarily replicate physics directly. Instead, they create

It means that the goal of glTF must be to parameterize the physical properties of real world materials and the interaction with light.
Not mimic the behavior of one or more given editing formats.
One of the reasons why glTF is so succesful is that we are agnostic and that the parameters used are a good destillation of physical properties.

Also of great importance - glTF is a `fileformat` - not a PBR model - you cannot compare glTF to existing PBR models.
Our job is to package the physical material and light properties into a dataset that can be distributet as a transmission format.

Reply:
I believe this thread was resolved during last week's PBR TSG meeting.